pkg/util/mount/mount_linux.go

```diff
@@ -1131,18 +1172,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
 	for _, seg := range segments {
 		currentPath = filepath.Join(currentPath, seg)
 		if !pathWithinBase(currentPath, base) {
-			return -1, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
+			return -1, defaultDeviceInfo, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
 		}
 
 		glog.V(5).Infof("Opening path %s", currentPath)
 		childFD, err = syscall.Openat(parentFD, seg, nofollowFlags, 0)
 		if err != nil {
-			return -1, fmt.Errorf("cannot open %s: %s", currentPath, err)
+			if err == syscall.ENXIO {
+				var deviceStat unix.Stat_t
+				deviceStatErr := unix.Fstatat(parentFD, seg, &deviceStat, unix.AT_SYMLINK_NOFOLLOW)
+				if deviceStatErr != nil {
+					return -1, defaultDeviceInfo, fmt.Errorf("Error getting stats for %s with %v", currentPath, deviceStatErr)
+				}
+
+				fileFmt := deviceStat.Mode & syscall.S_IFMT
+
+				if fileFmt == syscall.S_IFREG || fileFmt == syscall.S_IFDIR {
+					return -1, defaultDeviceInfo, fmt.Errorf("cannot open %s: %s", currentPath, err)
+				}
+
+				deviceInfo := fileDeviceInfo{
+					deviceNumber:  deviceStat.Dev,
+					iNodeNumber:   deviceStat.Ino,
+					specialDevice: true,
+				}
+				return -1, deviceInfo, nil
```

Comment:
returning here is incorrect if this is not the final segment (e.g. if the path was /foo/bar/baz and /foo/bar is a special file, we cannot descend to check the `baz` child and should fail)

Reply:
good point. should be easy to fix.